### DISCUSSION TITLE
Generating SDL and automatic resolvers (for graphql-codegen use)

### DISCUSSION THREAD

**Discussion by lenybernard** [TriPSs/nestjs-query#277] [Q&A] [answered]
Hi,

I'm facing a small issue. I just installed the library on an existing project which already had quite a few resolvers.

In this project, I need to generate the schema manually as explained here in the documentation: https://docs.nestjs.com/graphql/generating-sdl :

```typescript

async function generateSchema() {
  const app = await NestFactory.create(GraphQLSchemaBuilderModule);
  await app.init();

  const gqlSchemaFactory = app.get(GraphQLSchemaFactory);
  const schema = await gqlSchemaFactory.create([RecipesResolver]);
  console.log(printSchema(schema));
}
```

However, when I run the command, the generated schema does not take into account dynamic resolvers, i.e., those automatically generated by the syntax:

```typescript
NestjsQueryGraphQLModule.forFeature({
  imports: [NestjsQueryTypeOrmModule.forFeature([TodoItemEntity])],
  resolvers: [{ DTOClass: TodoItemDTO, EntityClass: TodoItemEntity }],
}),
```

Has anyone managed to do this? FYI, apart from this issue, everything works well in the app. I can access my resolvers, but the schema generated with generate-schema is incomplete, which is problematic because I need it to generate my TypeScript types file for the frontend (via graphql-codegen).

**Comment by TriPSs** [comment]
Could you give a bit more info?

For example, I also modify the schema file that is generated (copy a subset over to a other one) with no issues, I use:
```ts
GraphQLModule.forRootAsync({
  useFactory: (configService: ConfigService): MercuriusDriverConfig => ({
        autoSchemaFile: 'schema.main.gql',
        schemaTransforms: generateClientApiSchema
  })
})
```

```ts
export const generateClientApiSchema = (schema: GraphQLSchema): GraphQLSchema => {
    // Do stuff
    return schema
}
```

**Comment by lenybernard** [comment]
Thank you for your quick response, I pushed a repo to reproduce the issue https://github.com/lenybernard/nestjs-query-generate-sdl

My context is pretty basic :
| Description | File |
|--------|--------|
| some nestjs-query automatic resolvers on todo-item entity and dto | ️[src/todo-item/todo-item.module.ts#L11](https://github.com/lenybernard/nestjs-query-generate-sdl/blob/main/src/todo-item/todo-item.module.ts#L11) |
| a "native" Resolver based on a fake entity named `Author` |[src/todo-item/author.resolver.ts](https://github.com/lenybernard/nestjs-query-generate-sdl/blob/main/src/todo-item/author.resolver.ts) | 
| a generate-schema generating a schema in `dist/schema.graphql`| [src/generate-schema.ts](https://github.com/lenybernard/nestjs-query-generate-sdl/blob/main/src/generate-schema.ts) |

If you want to test, you can run : 

```bash
npm run generate-schema
```
Then, the generated schema file in `dist/schema.graphql` doesn't have CRUDResolver operations / documents but only those related to `AuthorResolver` :
```graphql
type Author {
  id: ID!
  name: String!
}

type Query {
  author(id: Int!): Author!
}
```

Thank you for your help.

**Reply by TriPSs** [reply]
I checked out your repo, but the output that is generated with `npm run generate-schema` is actually also what I would expect as you only provide `AuthorsResolver`.

Could you elaborate a bit more on your use case? Because I think you can also use the `schemaTransforms` approach I shared above.

**Reply by lenybernard** [reply]
Indeed, in the [generate-schema](https://github.com/lenybernard/nestjs-query-generate-sdl/blob/main/src/generate-schema.ts) file, I only put the AuthorsResolver because I don't know how to include automatic/generated resolvers.

The use case is simple : generating the graphql schema with every resolvers, static/classic like AuthorsResolver in my exemple and CrudResolvers like the one on TodoItemDTO. The process is ([correctly](https://docs.nestjs.com/graphql/generating-sdl)) documented in nestjs documentation for static/classic resolvers.

Automatic CRUDResolvers generated by the todo module's declaration :

```typescript
NestjsQueryGraphQLModule.forFeature({
      //...
      resolvers: [                                                    
        {
          EntityClass: TodoItemEntity,
          DTOClass: TodoItemDTO,
          CreateDTOClass: TodoItemCreateDTO,
        }, /* <- this will generate resolver that nestjs app will correctly enable 
                 but generate-schema.ts don't do it automatically */
      ],
    }),
```
Sorry but I don't understand your `schemaTransforms` approach, it may be the solution but how, should I use it in my generate schema script or how/where ?

**Reply by TriPSs** [reply]
Aah okay, I though you wanted to modify the generated schema file, you can also just add 

autoSchemaFile: 'schema.main.gql'

This will then generate the file with the defined name containing all the queries and mutations.

**Reply by lenybernard** [reply]
Thank you for your response even if I don't understand it 😄 and especially the `autoSchemaFile: 'schema.main.gql'` suggestion... 

Actually, I **do** want to modify the generated schema file from the `generate-schema.ts` script, because it's wrong (automatic resolvers and types are missing from it): 

```graphql
type Author {
  id: ID!
  name: String!
}

type Query {
  author(id: Int!): Author!
}
```

But in playground, I can download the following SDL (which is the expected) :
![image](https://github.com/user-attachments/assets/91cef456-53fb-4428-adba-2ee823fd1cbe)


```graphql
# Indicates exactly one field must be supplied and this field must not be `null`.
directive @oneOf on INPUT_OBJECT

type TodoItem {
  id: ID!
  title: String!
  completed: Boolean!
  created: DateTime!
  updated: DateTime!
}

# A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
scalar DateTime

type Author {
  id: ID!
  name: String!
}

type DeleteManyResponse {
  # The number of records deleted.
  deletedCount: Int!
}

type TodoItemDeleteResponse {
  id: ID
  title: String
  completed: Boolean
  created: DateTime
  updated: DateTime
}

type UpdateManyResponse {
  # The number of records updated.
  updatedCount: Int!
}

type TodoItemEdge {
  # The node containing the TodoItem
  node: TodoItem!

  # Cursor for this node.
  cursor: ConnectionCursor!
}

# Cursor for paging through collections
scalar ConnectionCursor

type PageInfo {
  # true if paging forward and there are more records.
  hasNextPage: Boolean

  # true if paging backwards and there are more records.
  hasPreviousPage: Boolean

  # The cursor of the first returned record.
  startCursor: ConnectionCursor

  # The cursor of the last returned record.
  endCursor: ConnectionCursor
}

type TodoItemConnection {
  # Paging information
  pageInfo: PageInfo!

  # Array of edges.
  edges: [TodoItemEdge!]!
}

type Query {
  author(id: Int!): Author!
  todoItem(
    # The id of the record to find.
    id: ID!
  ): TodoItem!
  todoItems(
    # Limit or page results.
    paging: CursorPaging! = { first: 10 }

    # Specify to filter the records returned.
    filter: TodoItemFilter! = {}

    # Specify to sort results.
    sorting: [TodoItemSort!]! = []
  ): TodoItemConnection!
}

input CursorPaging {
  # Paginate before opaque cursor
  before: ConnectionCursor

  # Paginate after opaque cursor
  after: ConnectionCursor

  # Paginate first
  first: Int

  # Paginate last
  last: Int
}

input TodoItemFilter {
  and: [TodoItemFilter!]
  or: [TodoItemFilter!]
  id: IDFilterComparison
  title: StringFieldComparison
  completed: BooleanFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

input BooleanFieldComparison {
  is: Boolean
  isNot: Boolean
}

input TodoItemSort {
  field: TodoItemSortFields!
  direction: SortDirection!
  nulls: SortNulls
}

enum TodoItemSortFields {
  id
  title
  completed
}

# Sort Directions
enum SortDirection {
  ASC
  DESC
}

# Sort Nulls Options
enum SortNulls {
  NULLS_FIRST
  NULLS_LAST
}

type Mutation {
  createOneTodoItem(input: CreateOneTodoItemInput!): TodoItem!
  createManyTodoItems(input: CreateManyTodoItemsInput!): [TodoItem!]!
  updateOneTodoItem(input: UpdateOneTodoItemInput!): TodoItem!
  updateManyTodoItems(input: UpdateManyTodoItemsInput!): UpdateManyResponse!
  deleteOneTodoItem(input: DeleteOneTodoItemInput!): TodoItemDeleteResponse!
  deleteManyTodoItems(input: DeleteManyTodoItemsInput!): DeleteManyResponse!
}

input CreateOneTodoItemInput {
  # The record to create
  todoItem: CreateTodoItem!
}

input CreateTodoItem {
  title: String!
  completed: Boolean!
}

input CreateManyTodoItemsInput {
  # Array of records to create
  todoItems: [CreateTodoItem!]!
}

input UpdateOneTodoItemInput {
  # The id of the record to update
  id: ID!

  # The update to apply.
  update: UpdateTodoItem!
}

input UpdateTodoItem {
  id: ID
  title: String
  completed: Boolean
  created: DateTime
  updated: DateTime
}

input UpdateManyTodoItemsInput {
  # Filter used to find fields to update
  filter: TodoItemUpdateFilter!

  # The update to apply to all records found using the filter
  update: UpdateTodoItem!
}

input TodoItemUpdateFilter {
  and: [TodoItemUpdateFilter!]
  or: [TodoItemUpdateFilter!]
  id: IDFilterComparison
  title: StringFieldComparison
  completed: BooleanFieldComparison
}

input DeleteOneTodoItemInput {
  # The id of the record to delete.
  id: ID!
}

input DeleteManyTodoItemsInput {
  # Filter to find records to delete
  filter: TodoItemDeleteFilter!
}

input TodoItemDeleteFilter {
  and: [TodoItemDeleteFilter!]
  or: [TodoItemDeleteFilter!]
  id: IDFilterComparison
  title: StringFieldComparison
  completed: BooleanFieldComparison
}
```

So nevermind, if you don't get it, I'll deal with it by committing the graphql codegen types file generated with valid SDL given directly by graphql server.
Thank you.

**Reply by lenybernard** [reply]
Oh this is better when adding `AppModule` in the list of resolvers:
https://github.com/lenybernard/nestjs-query-generate-sdl/commit/23c06e5c72e3dc1d14e623c80abc1c79f694faa3

```diff
import { NestFactory } from '@nestjs/core';
import { GraphQLSchemaBuilderModule, GraphQLSchemaFactory } from '@nestjs/graphql';
import { printSchema } from 'graphql/utilities';
import { promises as fs } from 'fs';
import * as path from 'node:path';
import { AuthorsResolver } from './todo-item/author.resolver';
+import { AppModule } from './app.module';

async function generateSchema() {
  const app = await NestFactory.create(GraphQLSchemaBuilderModule);
  await app.init();

  const gqlSchemaFactory = app.get(GraphQLSchemaFactory);
  const schema = await gqlSchemaFactory.create([
+    AppModule,
    AuthorsResolver
  ]);
  const distDir = path.join(__dirname, '../dist/');

  await fs.mkdir(distDir, { recursive: true });
  await fs.writeFile(path.join(distDir, 'schema.graphql'), printSchema(schema));
}

void generateSchema();
```

By doing this, the generated schema.graphl file contains more of things but it is still partial :

```graphql
type Author {
  id: ID!
  name: String!
}

type TodoItem {
  id: ID!
  title: String!
  completed: Boolean!
  created: DateTime!
  updated: DateTime!
}

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

type TodoItemEdge {
  """The node containing the TodoItem"""
  node: TodoItem!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type Query {
  author(id: Int!): Author!
}
```

For example, then, the TodoItems collection operation is missing and also the filters or Inputs...

**Reply by TriPSs** [reply, marked as answer]
> Thank you for your response even if I don't understand it 😄 and especially the autoSchemaFile: 'schema.main.gql' suggestion...

By doing this:
```ts
GraphQLModule.forRoot({
      autoSchemaFile: 'filename.gql'
})
```

A file name `filename.gql` will be generated that contains everything, this is also what I use to then generate my interfaces from (just like you).

**Reply by lenybernard** [reply]
Ok thank you, this solution is not based on the nestjs suggestion and I cannot just generate schema (without starting the app) but it is not worth and it works so I'll let it go like this.

For the record, I've added this conf and removed the generate-schema.ts as it doesn't generate full SDL
```diff
#app.module.ts
    }),
    GraphQLModule.forRoot<ApolloDriverConfig>({
      driver: ApolloDriver,
-      autoSchemaFile: true,
+      autoSchemaFile: 'src/schema.graphql',
    }),
    TodoItemModule,
  ],
```
[commit](https://github.com/lenybernard/nestjs-query-generate-sdl/commit/978a256f48c3a19e9ab5cbb03b6719a6717e217d)

Thank you !
